### PR TITLE
enhancement(filters): use a stricter bot token regex

### PR DESCRIPTION
**Pull request by onerandomusername**
Approval: https://canary.discord.com/channels/267624335836053506/635950537262759947/919203386660884560

Enhances the regex of the token remover to use the same regex that discord itself uses, with a slight modification. The `mfa` section was removed, but depending on an updated #1421, may be implemented. Additionally, the sections were grouped to keep working with the current code.

I kept the existing validation to keep false positives at a minimum. The current code checks the user resolves, the timestamp is valid, and the last section has at least 3 different characters.

As per @jb3, to implement:
- ~~#1421~~ now irrelevant.

Rejected:
- also check the upper bound of a token timestamp, to ensure it was in the past.

### CODE REVIEW

**Comment by onerandomusername** [comment]
I'll be fixing this commit history tonight

File: bot/exts/filters/token_remover.py [hunk 1]
```diff
@@ -78,6 +83,9 @@ async def on_message(self, msg: Message) -> None:
         if found_token:
             await self.take_action(msg, found_token)
```
Comment (Bluenix2) [inline bot/exts/filters/token_remover.py:84]: We could add a return after this - there's no point in checking for an MFA token.

**Review by Bluenix2 — APPROVED** [review]
Works as usual - I can't get access to an MFA token to test though. Thanks for this, I just have one comment that won't affect my review.

File: bot/exts/filters/token_remover.py (hunk 1 above)
Comment (onerandomusername) [reply bot/exts/filters/token_remover.py:84]: I think it is worth checking, or I should move checking for an mfa token before checking the other token format.

the reason for checking both was to ensure that a user who happened to share a user and bot token in the same message would not go undetected.

Comment (onerandomusername) [reply bot/exts/filters/token_remover.py:84]: Basically, with an mfa token those are to alert mods that someone sent it,  and don't let the user know why their messages was deleted intentionally.

however, I should probably go a little further with the implementation and possibly add some configuration as we may not want to ping moderators for mfa tokens

File: bot/exts/filters/token_remover.py [hunk 2]
```diff
@@ -167,7 +219,7 @@ def find_token_in_message(cls, msg: Message) -> t.Optional[Token]:
                 return token
 
         # No matching substring
-        return
+        return None
```
Comment (Akarys42) [inline bot/exts/filters/token_remover.py:222]: Why this change?

**Review by Akarys42 — APPROVED** [review]
LGTM, just a small style change

File: bot/exts/filters/token_remover.py (hunk 2 above)
Comment (onerandomusername) [reply bot/exts/filters/token_remover.py:222]: As the function can return a value, I wanted to make it concretely return None while I was here, rather than implicitly return None.

File: bot/exts/filters/token_remover.py (hunk 1 above)
Comment (Bluenix2) [reply bot/exts/filters/token_remover.py:84]: *Reading back my mails, I realize that it might be best for me to respond to this..*

Makes sense yeah, we'll keep it like this then.

File: bot/exts/filters/token_remover.py [hunk 3]
```diff
@@ -116,6 +124,50 @@ async def take_action(self, msg: Message, found_token: Token) -> None:
...
+
+        token = match.group()
+        # since the token was probably valid, we can now delete the message and ping the moderators.
+        # the user is not informed, given the reasoning for deleting the message.
+        with contextlib.suppress(NotFound):
+            await msg.delete()
+
+        log_message = (
+            f"Deleted mfa token sent by {msg.author} in {msg.channel}: "
+            f"{token[:4]}{'x' * len(token[3:-3])}{token[-3:]}"
```
Comment (wookie184) [inline bot/exts/filters/token_remover.py:154]:
```suggestion
            f"{token[:4]}{'x' * len(token[4:-3])}{token[-3:]}"
```
Just so the number of `x`s matches how many characters are missing :)

File: bot/exts/filters/token_remover.py [hunk 4]
```diff
@@ -116,6 +124,50 @@ async def take_action(self, msg: Message, found_token: Token) -> None:
...
+        Due to how the mfa tokens work, there is no user information supplied for this token.
+        """
+        # Ignore DMs; can't delete messages in there anyway.
+        if not msg.guild or msg.author.bot:
+            return
+
+        was_valid = False
+        for match in MFA_TOKEN_RE.finditer(msg.content):
+
+            if self.is_maybe_valid_hmac(match.group()):
```
Comment (wookie184) [inline bot/exts/filters/token_remover.py:140]: `match.group()` here will include `mfa.` so this will always return true, only the characters after should be passed here.

**Review by wookie184 — DISMISSED** [review]
Just a couple things, could you also make the message match the other token message a bit closer, particularly:
- having the user as a clickable mention and the ID in a codeblock afterwards
- having the channel as a clickable mention
- putting the censored token in a codeblock
- adding the pfp of the user that send the message as a thumbnail

File: bot/exts/filters/token_remover.py (hunk 3 above)
Comment (onerandomusername) [reply bot/exts/filters/token_remover.py]: https://github.com/python-discord/bot/pull/2006/commits/8fc503305cde2bf55a22a389bfdce1f7b01b6903

**Comment by wookie184** [comment]
This has been inactive for a while so I'll put it up for grabs.

**Comment by gustavwilliam** [comment]
> This has been inactive for a while so I'll put it up for grabs.

What's left to do? Addressing your comment and fixing the merge conflict?

**Comment by mbaruh** [comment]
> > This has been inactive for a while so I'll put it up for grabs.
> 
> What's left to do? Addressing your comment and fixing the merge conflict?

Yeah, merge main, match the output embed to the existing one, and fix the bug

**Comment by onerandomusername** [comment]
As far as I can tell the mfa code is no longer necessary as discord doesn't have mfa tokens like that now.

**Comment by onerandomusername** [comment]
There's also a bit of the matter that this regex is now out of date.

**Comment by wookie184** [comment]
> There's also a bit of the matter that this regex is now out of date. I don't know what the current token length is.

I can't recall any problems with false positives, even with the current regex, so we can afford to be very general with what we match. Could just check that the first and last parts are >= 10 and middle is >= 5 characters in length or something.

**Comment by Akarys42** [comment]
I agree with wookie, I think having extensive matching for the sake of even future-proofing is okay, considering additional checks are being performed

**Comment by wookie184** [comment]
> The only inaccurate part is the last section is too short as it is, but because of how the parsing works, that isn't an issue, i suppose
> 

If that's the case I think we should just have no upper limit on the length of the last section. Otherwise the code is somewhat misleading in how it works.

**Comment by mbaruh** [comment]
Hey @onerandomusername, thanks for your work so far. We needed to get this merged so went ahead and implemented the comments. Sorry the MFA part didn't end up being used 😅

File: bot/exts/filters/token_remover.py [hunk 5]
```diff
@@ -40,7 +40,7 @@
 # The HMAC isn't parsed further, but it's in the regex to ensure it at least exists in the string.
 # Each part only matches base64 URL-safe characters.
 # These regexes were taken from discord-developers, which are used by the client itself.
-TOKEN_RE = re.compile(r"([a-z0-9_-]{23,28})\.([a-z0-9_-]{6,7})\.([a-z0-9_-]{27})", re.IGNORECASE)
+TOKEN_RE = re.compile(r"([\w_-]{10,})\.([\w_-]{5,})\.([\w_-]{10,})", re.IGNORECASE)
```
Comment (onerandomusername) [inline bot/exts/filters/token_remover.py]: way too lenient, this makes *impossible* tokens match.

Comment (HassanAbouelela) [reply bot/exts/filters/token_remover.py]: Way too lenient is not defined by what it can match, but what it will match in practice, which so far has not been an issue. This PR made the regex a little more flexible in terms of false-positives, without increasing the possibility of false-negatives or creating something that has to be modified in the future.
